dstep/driver/Application.d

```diff
@@ -80,130 +87,159 @@ class Application : DStack.Application
+	return arguments.rawArgs[1 .. $] ~ argsToRestore;
     }
 
-    string[] extraArgs ()
+    string defaultOutputFilename ()
     {
-        return compiler.extraIncludePaths.map!(e => "-I" ~ e).toArray;
+	return Path.setExtension(arguments.argument.input, "d");
     }
 
-    string[] compilerArgs ()
+    
+    void handleLanguage (string language)
     {
-        return remainingArgs ~ extraArgs;
+	switch (language)
+	{
+	    case "c":
+	    case "c-header":
+	        this.language = Language.c;
+	    break;
+
+	    // Can't handle C++ yet
+	    //
+	    // case "c++":
+	    // case "c++-header":
+	    //     this.language = Language.cpp;
+	    // break;
+
+	    case "objective-c":
+	    case "objective-c-header":
+	        this.language = Language.objC;
+	    break;
+
+	    default:
+	        throw new DStepException(`Unrecognized language "` ~ language ~ `"`);
+	}
+
+	argsToRestore ~= "-x";
+	argsToRestore ~= language;
     }
-
-    bool handleDiagnostics ()
+    class ParseFile
     {
-        bool translate = true;
-
-        foreach (diag ; diagnostics)
-        {
-            auto severity = diag.severity;
-
-            with (CXDiagnosticSeverity)
-                if (translate)
-                    translate = !(severity == CXDiagnostic_Error || severity == CXDiagnostic_Fatal);
-
-            writeln(stderr, diag.format);
-        }
-
-        return translate;
+	private
+	{
+            string[] inputFiles;
+
+	    Index index;
+	    TranslationUnit translationUnit;
+	    DiagnosticVisitor diagnostics;
+
+	    Language language;
+	    string[] argsToRestore;
+	    Compiler compiler;
+	}
+
+	this(Language lang, string[] argTR)
+	{
+		language = lang;
+		argsToRestore = argTR.dup;
+	}
+
+	void startConversion (string file)
+	{
+		if (arguments["objective-c"])
+		    argsToRestore ~= "-ObjC";
+
+		index = Index(false, false);
+		translationUnit = TranslationUnit.parse(
+		    index, 
+		    file, 
+		    compilerArgs,
+		    compiler.extraHeaders);
+
+		// hope that the diagnostics below handle everything
+		// if (!translationUnit.isValid)
+		//     throw new DStepException("An unknown error occurred");
+
+		diagnostics = translationUnit.diagnostics;
+
+		scope (exit)
+		    clean;
+
+		if (handleDiagnostics && file.any)
+		{
+		    Translator.Options options;
+		    options.outputFile = arguments.output;
+		    options.language = language;
+
+		    auto translator = new Translator(file, translationUnit, options);
+		    translator.translate;
+		}
+	    }
+
+	private:
+
+	    void clean ()
+	    {
+		translationUnit.dispose;
+		index.dispose;
+	    }
+
+	    bool anyErrors ()
+	    {
+		return diagnostics.length > 0;
+	    }
+
+	    @property string[] remainingArgs ()
+	    {
+		return arguments.rawArgs[1 .. $] ~ argsToRestore;
+	    }
+
+	    string[] extraArgs ()
+	    {
+		return compiler.extraIncludePaths.map!(e => "-I" ~ e).toArray;
+	    }
+
+	    string[] compilerArgs ()
+	    {
+		return remainingArgs ~ extraArgs;
+	    }
```

Comment:
I don't think these three arguments related functions are necessary to have in this class. I think they will be the same for all instances.